docker/archive/tar.go

```diff
@@ -0,0 +1,210 @@
+	// file, it's necessary to store the end offset of the previous entry. This
+	// offset is rounded to the tarBlockSize.
+	var lastEntry int64
+
+	tr := tar.NewReader(tc.file)
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		// Add ourselves to the offset list.
+		tc.offsets[hdr.Name] = lastEntry
+
+		// Get current seek(2) offset and figure out where the end of our final
+		// chunk is for the next entry.
+		cur, err := tc.file.Seek(0, os.SEEK_CUR)
+		if err != nil {
+			return err
+		}
+
+		lastEntry = cur + hdr.Size
+
+		// Jump to the next block if this entry didn't round to a block.
+		if (cur+hdr.Size)%512 > 0 {
+			lastEntry += tarBlockSize - (lastEntry % tarBlockSize)
+		}
+	}
+
+	return nil
+}
+
+// jump will provide a tar.Header and tar.Reader for a given tar component by
+// using the pre-computed jump table (primed with .prime()) to simply seek to
+// the appropriate offset (avoiding the O(n) lookup time).
+func (tc *tarCache) jump(path string) (*tar.Header, *tar.Reader, error) {
+	off, ok := tc.offsets[path]
+	if !ok {
+		return nil, nil, fmt.Errorf("unknown entry: %s", path)
+	}
+
+	if _, err := tc.file.Seek(off, os.SEEK_SET); err != nil {
+		return nil, nil, err
+	}
+
+	tr := tar.NewReader(tc.file)
+	hdr, err := tr.Next()
+	if err != nil {
+		if err == io.EOF {
+			err = fmt.Errorf("jumped to end of archive: unknown entry: %s", path)
+		}
+		return nil, nil, err
+	}
+	return hdr, tr, nil
+}
+
+// recv streams the contents of the given tar component path, as well as
+// providing the size of the component (as reported by tar).
+func (tc *tarCache) recv(path string) (io.Reader, int64, error) {
+	hdr, tr, err := tc.jump(path)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	return tr, hdr.Size, nil
+}
+
+// read is identical to recv, except it will read the contents of the given tar
+// component path into a byte slice. Don't use this for big files.
+func (tc *tarCache) read(path string) ([]byte, error) {
+	r, _, err := tc.recv(path)
+	if err != nil {
+		return nil, err
+	}
+	bytes, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	return bytes, nil
+}
+
+// tarStat represents and augmented tar.Header that also contains the Digest of
+// the file.
+type tarStat struct {
+	*tar.Header
+	Digest string
+}
+
+// stat computes the map of Digests to tarStat inside the tar archive by
+// scanning it. This is different to the precomputed jump-table information
+// computed by .prime().
+func (tc *tarCache) stat() (map[string]*tarStat, error) {
+	// FIXME: We should probably cache these stats.
+	stats := map[string]*tarStat{}
+
+	for path := range tc.offsets {
+		hdr, tr, err := tc.jump(path)
+		if err != nil {
+			return nil, err
+		}
+
+		sum := sha256.New()
+		if _, err := io.Copy(sum, tr); err != nil {
+			return nil, err
+		}
+		digest := "sha256:" + hex.EncodeToString(sum.Sum(nil))
```

Comment:
Why is this necessary? Computing sha256 digests of hundreds of megabytes of data is pretty expensive.  We already know the DiffID values from the index→DiffID and index→file name mappings available in the tar manifest and image config (originally computed in `prepareLayerData`).